src/codegen/optimizer.lisp

```diff
@@ -316,6 +320,106 @@
+                                        (heuristic  #'aggressive-heuristic))
+  "Recursively inline function calls in `node`, using the environment
+`env` to look up the function bodies to substitute. The traversal
+keeps a call stack and will not continue past the specified
+`max-depth`. For recursive functions, traversal will not continue if
+the function already appears more than `max-unroll` times in the call
+stack. Otherwise, if a function's code can be found in `env` and the
+`heuristic` returns true, then the body will be inlined.
+
+Currently, this may break recursive data let expressions, since the
+codegen requires direct constructor calls."
+  (declare (type node           node)
+           (type tc:environment env)
+           (type (integer 0)    max-depth)
+           (type (integer 0)    max-unroll)
+           (type function       heuristic)
+           (values node &optional))
+  (flet ((try-inline (node call-stack)
+           (declare (type (or node-application node-direct-application) node)
+                    (type parser:identifier-list                        call-stack)
+                    (values (or node-let node-application node-direct-application) &optional))
+           (or
+            (alexandria:when-let*
+                ((name (node-rator-name node))
+                 (code (tc:lookup-code env name :no-error t))
+                 (_    (and (node-abstraction-p code)
+                            (funcall heuristic code)
+                            (<= (length call-stack)
+                                max-depth)
+                            (<= (count name call-stack)
+                                max-unroll)
+                            (= (length (node-abstraction-vars code))
+                               (length (node-rands node))))))
+              (multiple-value-bind (bindings subs)
+                  (loop :for var :in (node-abstraction-vars code)
+                        :for val :in (node-rands node)
+                        :for new-var := (gentemp (symbol-name var))
+                        :collect (cons new-var val) :into bindings
+                        :collect (make-ast-substitution
+                                  :from var
+                                  :to (make-node-variable
+                                       :type (node-type val)
+                                       :value new-var))
+                          :into subs
+                        :finally (return (values bindings subs)))
+                (make-node-let
+                 :type     (node-type node)
+                 :bindings bindings
+                 :subexpr  (funcall *traverse*
+                                    (rename-type-variables (apply-ast-substitution
+                                                            subs
+                                                            (node-abstraction-subexpr code)))
+                                    (cons name call-stack)))))
```

Comment:
This main inlining substitution is copied almost verbatim from https://github.com/coalton-lang/coalton/pull/1029 , just with type renaming and the stack